### PR TITLE
Added Bayesian Flows  :rocket: 

### PR DESCRIPTION
**Pull request by valsdav**
Bayesian Flows in zuko would be a very useful feature for uncertainty estimation with normalizing flows. 

This PR introduces Bayesian Flow in zuko with minimal changes:  `BayesianLinear` layers are provided as an alternative way of building the MLPs at the base of zuko transformations

All flows can be transformed to their Bayesian version by adding `bayesian=True` to the building arguments. 

An utility function `total_KL_divergence(model)` is added in the `zuko.utils` to compute the total KL divergence from all the bayesian layers in a model to be added to the loss. 

Let me know if you agree with the design of this feature! I added tests and a tutorial ;)

P.S.: CNF flows are not working for the moment but I haven't investigated why yet.

### CODE REVIEW

**Comment by francois-rozet** [comment]
Hi @valsdav 👋🏼 Thank you for your PR! I haven't had the time to take a look at it yet. I have a deadline by the end of May.

Until then, could you give a bit more context on this PR? Why is this feature necessary? What does it allow? Are there alternatives?

**Comment by valsdav** [comment]
HI @francois-rozet! Sure take your time.  I'm already using and testing more the PR in my group and so far so good!

This implementation is one of the way of encoding training uncertainty in a Normalizing Flow model. I'm interested in this application in the field of high energy physics, where estimating an uncertainty on the learnt probability density is crucial for downstream tasks. 

In particular bayesian networks can learn a posterior distribution over a model output making each network weight sampled from its own gaussian distribution. These networks, when properly trained with a KL divergence component in the loss, approximate well the statistical uncertainty from the limit amount of training dataset. 

In zuko the implementation is quite straightforward as it is enough to "make bayesian" all the linear models used inside the definition of the flow. At each forward call the linear NNs get sampled and the rest of the flow implementation stays the same. At inference time the user can call multiple time the forward function to build a distribution of the output prob. density. 

The idea of baysian NF is explored in https://arxiv.org/abs/2104.04543

Let me know if you prefer to have more references or more explanations!

**Comment by francois-rozet** [comment]
Hi @valsdav, thank you for remind me of the PR!

I have been taking a look this morning. Overall I think the implementation is good! However I have a few concerns regarding invertibility.

Transformations in normalizing flows are deterministic functions. Some (most) transformations call their underlying neural networks several times during a single forward. This is notably the case of auto-regressive transformations that call this several times

**Comment by francois-rozet** [comment]
> Would you recommend doing it with an utility function in zuko which is calling the flow after setting the necessary random tensors if "Bayesian" layers types are found?

What I imagine is to remove "Bayesian" layers, and instead provide functional helpers that operate on a standard (non-Bayesian) flow. Something like

```python
flow = zuko.flows.MAF(3)  # not Bayesian
prior = zuko.bayesian.diagonal_gaussian_prior(flow.parameters())  # special object

optim = torch.optim.AdamW(prior.parameters(), lr=1e-3)

for x in train_loader:
    phi = prior().sample()
    with zuko.bayesian.parameterize(flow, phi):
        log_p = flow().log_prob(x)
    loss = -log_p.mean()
    loss.backward()
    
    optim.step()
    optim.zero_grad()

with torch.no_grad():
    phi = prior().sample()
    with zuko.bayesian.parameterize(flow, phi):
        x = flow().sample()
```

**Comment by francois-rozet** [comment]
> In practice this prior object would store the mean and std of the weights of Linear layers.

Yes exactly! Maybe of the biases as well, although I don't know if this is common.

> but then it makes more difficult on the user side to store properly the "prior" parameters alongside the flow model itself.

Is your concern about saving and loading the models to disk?

> What is we wrap this "prior+flow" object in a new flow object? So that we can store them together and provide the same user interface?

This is another option to consider. We should however ensure that the user understands what this new "prior + flow" object represents and how to use it "like other Zuko flows".

I usually like when everything is very explicit for the user. In my snippet it is clear that what is optimized is the prior, not the flow.

**Comment by valsdav** [comment]
Ok thanks for the discussion. I will give a try to this new implementation and come back to you :)

File: zuko/bayesian.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        # print(f"[BayesianFlow.sample] Starting sample context (training={self.training})")
+
+        if self.training:
+            # Use local reparameterization trick for training (more efficient)
+            with self._reparameterization_context():
+                yield self.base
+        else:
+            # Use parameter sampling for evaluation
+            sampled_params = self._sample_params()
+            proxy = self._create_sampled_proxy(sampled_params)
```
[the remark this answers is not in the document]
Comment (valsdav) [reply zuko/bayesian.py]: Hi @francois-rozet! I tried using the `functional_call` but I realized that it was not working as the Transformations are calling again the forward method after the first call to `forward`.    So in my case the reparametrization was not actually used to change the parameters of the flow while calling rsample or log_prob. 

I can try using directly the `torch.nn.utils.stateless._reparametrize_module` method as it does not only change the parametrization during the first call.

File: zuko/bayesian.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            b_var = torch.exp(b_logvar)
+            if isinstance(module, MaskedLinear):
+                # Apply mask to bias variance - use mask's output dimension
+                masked_b_var = module.mask.any(dim=1).float() * b_var
+                var_out = var_out + masked_b_var
+            else:
+                var_out = var_out + b_var
+
+        # Sample output using reparameterization trick
+        result = mu_out + var_out.sqrt() * torch.randn_like(mu_out)
```
[the remark this answers is not in the document]
Comment (valsdav) [reply zuko/bayesian.py:284]: right! I missed this thanks

File: zuko/bayesian.py [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+import warnings
+
+from contextlib import contextmanager
+
+from .nn import Linear, MaskedLinear, linear
+
+
+def _softclip(x, extreme_val=11.0):
+    """Soft clipping function to keep x in [extreme_val, extreme_val]."""
+    return x / (1 + (x.square() / (extreme_val**2))).sqrt()
```
Comment (francois-rozet) [inline zuko/bayesian.py:22]:
```suggestion
def _softclip(x, bound=11.0):
    """Soft clipping function to keep x in [-bound, bound]."""
    return x * (1 + (x / bound).square()).rsqrt()
```

Any reason for using 11.0?

File: zuko/bayesian.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+        # bayesian_layers: optional list of module names (dotted, e.g. "layer1.linear")
+        # If None, all Linear/MaskedLinear modules are treated as Bayesian (existing behavior).
+        # Store both dotted and underscore-safe names for easy matching.
+        if bayesian_layers is None:
+            self._requested_bayesian = None
+        else:
+            safe = set()
+            for n in bayesian_layers:
+                safe.add(n)
+            self._requested_bayesian = safe
```
Comment (francois-rozet) [inline zuko/bayesian.py:45]:
```suggestion
            self._requested_bayesian = set(bayesian_layers)
```

File: zuko/bayesian.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+                    self.bias_logvars[safe_name + "_bias"] = nn.Parameter(
+                        torch.full_like(module.bias, init_logvar)
+                    )
+
+        # Initialize them properly
+        self._reset_bayesian_parameters(init_logvar)
+
+        # Validate requested bayesian layer names and warn if any weren't found
+        if self._requested_bayesian is not None:
+            missing = set(self._requested_bayesian) - matched
```
Comment (francois-rozet) [inline zuko/bayesian.py:89]:
```suggestion
            missing = self._requested_bayesian - matched
```

File: zuko/bayesian.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+                        def reparameterized_forward(x):
+                            return self._apply_local_reparameterization_eval(
+                                x, module_name, original_module
+                            )
+
+                    return reparameterized_forward
+
+                # Replace the forward method using setattr on instance to avoid mypy complaints
+                module.forward = make_reparameterized_forward(name, module)
```
Comment (francois-rozet) [inline zuko/bayesian.py:214]: Instead of using a closure, you can use `functools.partial`:

```suggestion
                    if self.training:
                        model.forward = partial(
                            self._apply_local_reparameterization_training,
                            name=name,
                            module=module,
                        )
                    else:
                        model.forward = partial(
                            self._apply_local_reparameterization_eval,
                            name=name,
                            module=module,
                        )
```

File: zuko/bayesian.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (francois-rozet) [reply zuko/bayesian.py:276]: I agree that this would be the simplest (and what I had imagined). However, I don't think this is compatible with the training trick for the variance. That is why I mentioned using a fixed random seed, such that you can still perform the trick, but the call becomes deterministic within a context.

[the remark this answers is not in the document]
Comment (valsdav) [reply zuko/bayesian.py:276]: sorry getting confused here. This is already how it would work without the training trick, just by using again the `torch.nn.utils.stateless._reparametrize_module` strategy.   

Now for the training trick, I don't want to resample the weights but apply the random smearing on the average output. In this case I don't know how to do it as I would need to store the random values for each layer and re-apply them the second time forward get called the same.  Setting the same seed would not work as the generator would get new numbers

File: zuko/bayesian.py [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    "BayesianModel",
+]
+
+import copy
+import math
+import torch
+import torch.nn as nn
+import warnings
+
+from contextlib import contextmanager
```
Comment (francois-rozet) [inline zuko/bayesian.py:17]:
```suggestion
__all__ = [
    "BayesianModel",
]

import copy
import math
import torch
import torch.nn as nn
import warnings

from contextlib import contextmanager
from functools import partial
```

File: zuko/bayesian.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+                warnings.warn(
+                    f"BayesianModel: requested bayesian_layers not found in base model: {sorted(missing)}",
+                    UserWarning,
+                    stacklevel=2,
+                )
+
+    def reset_generator(self, device="cpu"):
+        """Rebuild generator from stored seed (e.g. after loading state_dict).
+        This is call each time"""
+        self.generator = torch.Generator(device=device).manual_seed(int(self.rng_seed))
```
Comment (francois-rozet) [inline zuko/bayesian.py:103]: This isn't necessary IMO. Torch modules rarely handle their own randomness.

File: zuko/bayesian.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+        original_forwards = {}
+
+        # Sample a new seed for each module call
+        for name, module in self.base.named_modules():
+            if isinstance(module, (Linear, MaskedLinear)):
+                # Skip non-selected layers when a subset was provided
+                if self._requested_bayesian is not None and name not in self._requested_bayesian:
+                    continue
+
+                seed = int(torch.randint(0, 2**31 - 1, (1,), generator=self.generator).item())
```
Comment (francois-rozet) [inline zuko/bayesian.py:207]: You can use the global random state here. 

```suggestion
                seed = torch.randint(0, 2**31 - 1, ()).item()
```

File: zuko/bayesian.py [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+                        module=module,
+                        seed=seed,
+                    )
+                else:
+                    module.forward = partial(
+                        self._apply_local_reparameterization_eval,
+                        name=name,
+                        module=module,
+                        seed=seed,
+                    )
```
Comment (francois-rozet) [inline zuko/bayesian.py:223]: This does not exist anymore?

File: zuko/bayesian.py [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+
+        # Add bias variance if present
+        if module.bias is not None:
+            b_logvar = _softclip(self.bias_logvars[safe_name + "_bias"], 11)
+            b_var = torch.exp(b_logvar)
+            if isinstance(module, MaskedLinear):
+                # Apply mask to bias variance - use mask's output dimension
+                masked_b_var = module.mask.any(dim=1).float() * b_var
+                var_out = var_out + masked_b_var
+            else:
```
[the remark this answers is not in the document]
Comment (valsdav) [reply zuko/bayesian.py:284]: right, this is not needed, only the inputs are masked

File: zuko/bayesian.py [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    """Soft clipping function to keep x in [-bound, bound]."""
+    return x * (1 + (x / bound).square()).rsqrt()
+
+
+class BayesianModel(nn.Module):
+    def __init__(
+        self,
+        base: nn.Module,
+        init_logvar: float = -9.0,
+        seed: int = 42,
```
Comment (francois-rozet) [inline zuko/bayesian.py:30]:
```suggestion
```

**Comment by francois-rozet** [comment]
Yes, exactly, the local reparam trick assumes that the output is `x * (w + noise)` but here we have `x * abs(w + noise)`. We need a special trick that assumes `x * (abs(w) + abs(noise))` or such that the noise is positive (e.g. $\chi^2$ distribution).

**Comment by francois-rozet** [comment]
@valsdav I have started editing the PR on my side, to clean up the implementation, add docstrings, and improve the tests. Please do not commit.

**Comment by valsdav** [comment]
Hi @francois-rozet I did some more tests on the latest version and we have a regression somewhere. The flows is not learning anymore during the training. Did I messed up the gradient propagation? Having a closer look

**Comment by francois-rozet** [comment]
Hi @valsdav, I went over the code. I have refactored it to make it future proof (e.g. `self.means` instead of `self.weight_means`). Instead of replacing `.` with `_`, I replace it with `-`. This prevents potential name clashes (e.g. `my_layer.weight` and `my.layer.weight`). I have also added docstrings.

The bug with the `GF` was due to the initialization of the weights, which were too large. I prefer to fallback to the initial base model weights.

Finally, `sample_model` should never be used to train as `load_state_dict` breaks the gradients to the Bayesian model parameters.

**Comment by valsdav** [comment]
Hi @francois-rozet looks great to me! The code has become much better :) 

Thanks a lot, +1 for merging on my side, we would start using this new version right away.

**Comment by valsdav** [comment]
Excluding with a patter is neat, but the usecase for me would be to make bayesian only the last layers of a deeper net (a common practice to avoid making large bayesian net hard to train). In this case would would need to exclude all the other layers.   I can try to implement in a similar way an "include_only" option. 

I will have a look again at the notebook and update it in the next days :+1:

**Comment by francois-rozet** [comment]
Oh it makes sense for the include 🤔 I think we can have both an include and exclude list then. We first include `if any(name.startswith(prefix) for prefix in include_modules)` and then exclude `if any(name.startswith(prefix) for prefix in exclude_modules)`.

If we want to allow pattern matching, I think wildcards `*` would be nice (e.g. `**.hyper.*.*` matches `transform.hyper.1.weight`).

**Comment by francois-rozet** [comment]
Hi @valsdav, after fixing the tests, I noticed more bugs with `NAF` and `UNAF`. This is due to the use of `self.parameters()` in `MNN` and `UMNN`, which breaks with `_reparameterize_module` **and** the local reparametrization trick (in subtle, different ways). I have decided to drop official support for these flows.

I have updated the tutorial to be more consistent with the other ones. I am ready to merge.
